### ISSUE TITLE
Type and term equality is still broken

### ISSUE THREAD

**Issue by effectfully** [input-output-hk/plutus#1611]
These two types are considered equal by the current machinery:

```
forall (x_3 :: *) (x_3 :: *). x_3 -> x_3
forall (y_4 :: *) (z_5 :: *). y_4 -> z_5
```

I've been dreaming of removing the current bloated equality checking machinery and replacing it with a more sensible one for quite a while. Now I have a good reason for doing that.

**Comment by effectfully** [comment]
Okay, here are the options:

0. give up on equality modulo alpha. I don't think it's a good idea

1. fix the bug simply by using [`Bimap`](http://hackage.haskell.org/package/bimap-0.4.0/docs/Data-Bimap.html#t:Bimap)s instead of `Map`s. Two variables are considered equal only when they are bound by the same binder and if one of these variables (or both of them) was bound previously, we shadow both the existing bindings. This is exactly the way [`Bimap.insert`](http://hackage.haskell.org/package/bimap-0.4.0/docs/Data-Bimap.html#v:insert) works. This way we can ensure that variables on either side are scoped properly (the current machinery only does that for the right side)

2. but I'd really rather rewrite the entire machinery to distinguish between scopes in a type-directed manner (in the style of the renamer) and for this we need a `bi` version of `UniqueMap`, but its underlying type is `IntMap` which doesn't have a `bi` version. But `Bimap` is defined as

```haskell
data Bimap a b = MkBimap !(M.Map a b) !(M.Map b a)
```

and we can easily define `IntBimap` in the same manner. So we can define `UniqueBimap` and use the same approach as is currently implemented, but in a nice type-directed manner

3. rename types/terms and then use the straightforward equality. Downsides:

- really inefficient: we need to traverse both the terms looking for free variables to initialize the renamer (to prevent capturing of free variables; the renamer does that behind the scenes), then to rename the terms and construct their renamed versions in memory and only then perform the check. This is especially wasteful when the terms are not equal and instead of just immediately returning `false` we first rename the terms
- kind of weird to make equality depend on the renamer

4. rename and check for equality simultaneously. Still need a separate pass to handle free variables. I think, this is better than 3, but more code and still not good enough 

I'm leaning towards 2. Thoughts?

**Comment by michaelpj** [comment]
I'd say 1 or 2 seems fine. I suspect that implementing 1 is naturally part of the refactoring towards implementing 2, so I'd do that first, and then try extending it.

I agree that rename then equate in its various forms seems inelegant and inefficient.

**Comment by effectfully** [comment]
> Two variables are considered equal only when they are bound by the same binder and if one of these variables (or both of them) was bound previously, we shadow both the existing bindings.

But this doesn't handle free variables. We could say that two variables are equal either if they were introduced by the same or if they already have equal uniques, but that also would be wrong, because by this logic

```
forall x_0. x_0
forall x_1. x_0
```

are also considered equal. Clearly variables can only be equal when they are both either bound or free and we know how to handle both the cases separarely (bound -- via `Bimap`, free -- via straightforward equality), but keeping track of scopes on each side in addition to track it simultaneously is quite annoying.

**Comment by effectfully** [comment]
> rename and check for equality simultaneously. Still need a separate pass to handle free variables. I think, this is better than 3, but more code and still is not good enough

Actually, no, we can do that in one pass easily:

- keep track of scopes on each side
- on encountering twin binders, record that each of the possibly distinct variables maps to the same fresh variable
- when checking two variables for equality, look them up in the current scope
- + if both are in scope, then check that they map to the same variable
- + if both are not in scope, then check their direct equality
- + if one is in scope and the other one is not, then return `False`

This way we never check for equality a bound variable and a free variable and thus can map bound variables to uniques already reserved by free variables.

I'll implement this.

**Comment by michaelpj** [comment]
Hm, would it be simpler to just rename first and then do the simple check (although: what about free variables?)? I'm not too worried about performance of equality, to be honest, I don't think anything much depends on it.

**Comment by effectfully** [comment]
> Hm, would it be simpler to just rename first and then do the simple check (although: what about free variables?)?

The renamer handles free variables, so it's an option.

> I'm not too worried about performance of equality, to be honest, I don't think anything much depends on it.

Tests depend and it's nice to have faster tests.

**Comment by michaelpj** [comment]
The tests are not especially slow, I think!

**Comment by effectfully** [comment]
> Hm, would it be simpler to just rename first and then do the simple check (although: what about free variables?)?

The simple check is not just direct syntactic equality. Two renamed terms will still have distinct uniques and we'll have to track variables bound by twin binders.

I'd rather directly handle scopes, then first rename, then rely on global uniqueness and write even more code in the end.

**Comment by effectfully** [comment]
> on encountering twin binders, record that each of the possibly distinct variables maps to the same fresh variable

We can in fact map them to each other like with the `Bimap` solution except this problem:

> Clearly variables can only be equal when they are both either bound or free and we know how to handle both the cases separarely (bound -- via Bimap, free -- via straightforward equality), but keeping track of scopes on each side in addition to tracking them simultaneously is quite annoying.

is solved via 

> - when checking two variables for equality, look them up in the current scope
>   + if both are in scope, then check that they map to the same variable
>   + if both are not in scope, then check their direct equality
>   + if one is in scope and the other one is not, then return False

**Comment by effectfully** [comment]
Fixed by #1655.
